### PR TITLE
Ud/project improv

### PR DESCRIPTION
**Pull request by Udayan12167**
**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

File: data/projects/coala_cli.md [hunk 1]
```diff
@@ -1,24 +1,53 @@
-[cEP-0004](https://github.com/coala/cEPs/pull/29/files) was meant to improve
-the usability of coala. It is a very important proposal that never got the time
-and love it deserved.
+coala has come a long way from it's early days in terms of it's usability. But
+all the changes that have been made are improvisations. A proper and planned
+overhaul of the user interface would help coala become much more user
+friendly and intuitive. At it's core any software is as good as it's
+interface with the user and this project aims to improve that.
+
+In this project we are going to focus on the command line interface(CLI) of
```
Comment (NiklasMM) [inline data/projects/coala_cli.md]: A space between interface and (CLI)?

File: data/projects/coala_cli.md [hunk 2]
```diff
@@ -1,24 +1,53 @@
...
+coala. The specific areas and points to focus on are:
+
+1. Remove most of the text from the action selection screen.
+2. Use letters instead of numbers for selecting actions.
+3. Allow chaining actions in a single input.
+4. Add a 'select action for all results' operator.
+5. Make ApplyPatchAction log the applied patch.
```
Comment (NiklasMM) [inline data/projects/coala_cli.md]: Add 's around ApplyPatchAction

File: data/projects/coala_cli.md [hunk 3]
```diff
@@ -1,24 +1,53 @@
...
+6. Reprompt after running actions only if the action says so.
+7. Set 'Do nothing' as the default action.
+
+The end goal of the project is to make coala easier to use and hence the
```
Comment (NiklasMM) [inline data/projects/coala_cli.md]: "and hence" -> ", hence" :)

File: data/projects/coala_cli.md [hunk 4]
```diff
@@ -1,24 +1,53 @@
...
+project is going to be heavy on mockups and feedback from the community.
```
Comment (NiklasMM) [inline data/projects/coala_cli.md]: "be heavy" -> "rely heavy"

File: data/projects/coala_cli.md [hunk 5]
```diff
@@ -1,24 +1,53 @@
...
+Also, in addition to the above changes the applicant is encouraged to come up
+with his/her own ideas to improve the CLI.
```
Comment (NiklasMM) [inline data/projects/coala_cli.md]: his/her -> their

File: data/projects/coala_cli.md [hunk 6]
```diff
@@ -1,24 +1,53 @@
...
+CLI.
+* The participant has a clear idea on how all parts of the cEP can be
+implemented.
 
-##### GSOC 2017 COMMUNITY BONDING
+##### CODING PHASE 1
 
-* The cEP has been updated, finished and merged.
-* The participant has a clear idea on how all parts of the cEP can be implemented.
+* Implemented atleast 3 improvements from the above list plus whatever has
```
Comment (NiklasMM) [inline data/projects/coala_cli.md]: space missing in "at least"

File: data/projects/coala_cli.md [hunk 7]
```diff
@@ -1,24 +1,53 @@
...
+been proposed by the participant.
+* Thorough testing of the implemented improvements is done and they are merged.
+* Documentation using illustrations of the improvements is merged.
 
-##### GSOC 2017 MIDTERM
+##### CODING PHASE 2
 
-* At least half of the points given in the cEP are done, merged, tested and properly documented.
+* Implemented ,documented and tested atleast 3 more improvements with coding
```
Comment (NiklasMM) [inline data/projects/coala_cli.md]: "Implemented ,documented" -> "Implemented, documented"

File: data/projects/coala_cli.md [hunk 8]
```diff
@@ -1,24 +1,53 @@
...
+finished and merged.
 
-##### GSOC 2017 FINAL
+##### CODING PHASE 3
 
-* All of the points given in the cEP are done, merged, tested and properly documented.
+* Implemented ,documented and tested all improvements and merged them.
```
Comment (NiklasMM) [inline data/projects/coala_cli.md]: "Implemented ,documented" -> "Implemented, documented"

File: data/projects/debug_profile.md [hunk 9]
```diff
@@ -1,26 +1,49 @@
-In the development lifecycle of a Bear it needs to be debugged and profiled
-and coala should provide a comprehensive environment
-for this. In this project the participant shall develop a system to easily
+A famous programming quote reads "Debugging is twice as hard as writing code
+in the first place". Thus even though writing code can be as trivial as
+opening up your favorite text editor and crunching some keys we need tools to
+help us debug our code. An essential toolset that coala lacks is the ability
+to debug it's bears i.e the analysis procedures. In the development lifecycle
```
Comment (NiklasMM) [inline data/projects/debug_profile.md]: [Comma before "i.e."](https://jakubmarian.com/comma-after-i-e-and-e-g/)

File: data/projects/debug_profile.md [hunk 10]
```diff
@@ -1,26 +1,49 @@
...
+of a Bear it needs to be debugged and profiled and coala should provide a
```
Comment (NiklasMM) [inline data/projects/debug_profile.md]: Please be consistent about the capitalization of "bear". Here it's upper case and in line 5 it's lower case.

File: data/projects/debug_profile.md [hunk 11]
```diff
@@ -1,26 +1,49 @@
...
+comprehensive environment for this.
+
+In this project the participant shall develop a system to easily
 step into the analysis code (likely using the Pdb module already integrated
-into Python) and step out as soon as the analysis finished.
+into Python) and step out as soon as the analysis finished. The following are desireable goals for such a toolset implementaion:
```
Comment (NiklasMM) [inline data/projects/debug_profile.md]: The following lists defines features rather than implementation details, therefore "toolset implementaion" -> "toolset's features"

File: data/projects/debug_profile.md [hunk 12]
```diff
@@ -1,26 +1,49 @@
...
+
+1. Tha ability to jump directly into bear code and move through it using a debugger interface such as pdb.
```
Comment (NiklasMM) [inline data/projects/debug_profile.md]: The

File: data/projects/improve_diff_handling.md [hunk 13]
```diff
@@ -1,24 +1,41 @@
...
-for different possible spelling corrections.
-
-In addition bears should be able to offer templated patches, e.g. to offer
+The advantage of using coala is that bears offer automatically generated diffs
+to fix the problems they detect. This mechanism works smoothly as of now and
+even applies the proposed diff automatically to the problematic file if the
+user so desires. This is very helpful to the users but can be improved by
+giving coala the ability to:
+
+1. Offer multiple diffs i.e. multiple different fixes to the same issue.
```
Comment (NiklasMM) [inline data/projects/improve_diff_handling.md]: comma before "i.e."

File: data/projects/improve_diff_handling.md [hunk 14]
```diff
@@ -1,24 +1,41 @@
...
+Giving the user a choice to pick the most desireable one.
+2. Incorporate input from the user into the diffs offered by using a
+templating mechanism.
+
+The above improvements can be used in these sample situations. For e.g. when
```
Comment (NiklasMM) [inline data/projects/improve_diff_handling.md]: "For e.g." doesn't work. Either "for example" or "e.g."

File: data/projects/coala_cli.md (hunk 4 above)
Comment (RohanVB) [inline data/projects/coala_cli.md]: "rely heavily upon" sounds better imo

File: data/projects/coala_cli.md [hunk 15]
```diff
@@ -1,24 +1,53 @@
-[cEP-0004](https://github.com/coala/cEPs/pull/29/files) was meant to improve
-the usability of coala. It is a very important proposal that never got the time
-and love it deserved.
+coala has come a long way from it's early days in terms of it's usability. But
```
Comment (RohanVB) [inline data/projects/coala_cli.md]: Maybe its not right to start with a conjunction here.. "usability. But" -> "usability, although"
also, "it's" - > "its"

File: data/projects/debug_profile.md [hunk 16]
```diff
@@ -1,26 +1,49 @@
...
 
-##### GSOC 2017 FINAL
+##### CODING PHASE 3
 
-* The profile option to CLI is added and merged.
+* The profile option to CLI is merged.
 * The ability to inspect results produced by the bear is implemented.
 * The documentation about profiling is written.
+* A demo of how to use the profiling and debuggin is included in the
+documentation.
```
Comment (RohanVB) [inline data/projects/debug_profile.md]: double space here to make the second line inline with the above line

File: data/projects/improve_diff_handling.md [hunk 17]
```diff
@@ -1,24 +1,41 @@
...
 * The user is able to choose between multiple patches.
 * The new API is fully documented on <https://api.coala.io/>.
-* At least one bear exists that makes use of this.
+* Proposal and implementation of the bear using multiple patches.
+
+##### CODING PHASE 2
+
+* Implementation of the multiple diff bear is completed and it is merged.
+* The implementation for templated patches is nearing completion and it is
+being tested.
```
Comment (RohanVB) [inline data/projects/improve_diff_handling.md]: double space here to make the second line inline with the above line

File: data/projects/coala_cli.md [hunk 18]
```diff
@@ -1,24 +1,53 @@
...
 
-Ideally a candidate comes up with own ideas that are not sketched in the cEP
-yet!
+##### PREPARATION/BONDING
 
-#### MILESTONES
+* The [cEP](https://coala.io/cep) has been updated, finished and merged with
+the help of mockups and community feedback.
+* Preferably a survey is done on the importance of each proposed change to the
+CLI.
```
Comment (RohanVB) [inline data/projects/coala_cli.md]: double space here to make the second line inline with the above line

File: data/projects/coala_cli.md [hunk 19]
```diff
@@ -1,24 +1,53 @@
...
+* The participant has a clear idea on how all parts of the cEP can be
+implemented.
```
Comment (RohanVB) [inline data/projects/coala_cli.md]: double space here to make the second line inline with the above line

File: data/projects/coala_cli.md [hunk 20]
```diff
@@ -1,24 +1,53 @@
...
-defined in it.
+#### MILESTONES
 
-Ideally a candidate comes up with own ideas that are not sketched in the cEP
-yet!
+##### PREPARATION/BONDING
 
-#### MILESTONES
+* The [cEP](https://coala.io/cep) has been updated, finished and merged with
+the help of mockups and community feedback.
```
Comment (RohanVB) [inline data/projects/coala_cli.md]: double space here to make the second line inline with the above line

File: data/projects/improve_diff_handling.md [hunk 21]
```diff
@@ -1,24 +1,41 @@
...
 
 ##### GSOC 2017 FINAL
 
-* The user is able to apply templated patches.
+* Templated patches is merged and fully tested.
```
Comment (RohanVB) [inline data/projects/improve_diff_handling.md]: "is" -> "are"

File: data/projects/improve_diff_handling.md [hunk 22]
```diff
@@ -1,24 +1,41 @@
...
+##### CODING PHASE 1
 
 * The user is able to choose between multiple patches.
 * The new API is fully documented on <https://api.coala.io/>.
-* At least one bear exists that makes use of this.
+* Proposal and implementation of the bear using multiple patches.
+
+##### CODING PHASE 2
+
+* Implementation of the multiple diff bear is completed and it is merged.
```
Comment (RohanVB) [inline data/projects/improve_diff_handling.md]: quotes around multiple diff bear?

File: data/projects/improve_diff_handling.md [hunk 23]
```diff
@@ -1,24 +1,41 @@
...
+user so desires. This is very helpful to the users but can be improved by
+giving coala the ability to:
+
+1. Offer multiple diffs i.e. multiple different fixes to the same issue.
+Giving the user a choice to pick the most desireable one.
+2. Incorporate input from the user into the diffs offered by using a
+templating mechanism.
+
+The above improvements can be used in these sample situations. For e.g. when
+suggesting natural language corrections it can offer multiple spelling
```
Comment (RohanVB) [inline data/projects/improve_diff_handling.md]: "corrections it" -> "corrections, it"

File: data/projects/improve_diff_handling.md (hunk 22 above)
Comment (Udayan12167) [inline data/projects/improve_diff_handling.md]: Why are the quotes needed?

Comment (RohanVB) [inline data/projects/improve_diff_handling.md]: Is 'multiple diff bear' a new bear to be implemented? I must've not understood this sentence if it's not.

Comment (Udayan12167) [inline data/projects/improve_diff_handling.md]: Yes it's a new bear

File: data/projects/coala_cli.md [hunk 24]
```diff
@@ -1,24 +1,53 @@
-[cEP-0004](https://github.com/coala/cEPs/pull/29/files) was meant to improve
-the usability of coala. It is a very important proposal that never got the time
-and love it deserved.
+coala has come a long way from its early days in terms of it's usability,
```
Comment (RohanVB) [inline data/projects/coala_cli.md]: "it's" -> "its"

File: data/projects/coala_cli.md [hunk 25]
```diff
@@ -1,24 +1,53 @@
...
+athough all the changes that have been made are improvisations. A proper and
+planned overhaul of the user interface would help coala become much more user
+friendly and intuitive. At it's core any software is as good as it's interface
```
Comment (RohanVB) [inline data/projects/coala_cli.md]: "it's" -> "its" (twice)

File: data/projects/debug_profile.md [hunk 26]
```diff
@@ -1,26 +1,50 @@
-In the development lifecycle of a Bear it needs to be debugged and profiled
-and coala should provide a comprehensive environment
-for this. In this project the participant shall develop a system to easily
+A famous programming quote reads "Debugging is twice as hard as writing code
+in the first place". Thus even though writing code can be as trivial as
+opening up your favorite text editor and crunching some keys we need tools to
+help us debug our code. An essential toolset that coala lacks is the ability
+to debug it's Bears, i.e. the analysis procedures. In the development lifecycle
```
Comment (RohanVB) [inline data/projects/debug_profile.md]: "it's" -> "its"

File: data/projects/debug_profile.md [hunk 27]
```diff
@@ -1,26 +1,50 @@
-In the development lifecycle of a Bear it needs to be debugged and profiled
-and coala should provide a comprehensive environment
-for this. In this project the participant shall develop a system to easily
+A famous programming quote reads "Debugging is twice as hard as writing code
+in the first place". Thus even though writing code can be as trivial as
+opening up your favorite text editor and crunching some keys we need tools to
+help us debug our code. An essential toolset that coala lacks is the ability
+to debug its Bears, i.e. the analysis procedures. In the development lifecycle
+of a Bear it needs to be debugged and profiled and coala should provide a
+comprehensive environment for this.
```
Comment (sims1253) [inline data/projects/debug_profile.md]: Ẁhy do we need this?

File: data/projects/debug_profile.md [hunk 28]
```diff
@@ -1,26 +1,50 @@
...
+performant Bears through the profiling interface. It also makes Bear
+developement easier through the debugging interface. Also, it gives the added
+benefit of better quality Bears being contributed to the community.
 
 #### Milestones
 
-##### GSOC 2017 COMMUNITY BONDING
+##### PREPARATION/BONDING
 
 * The applicant has made a [cEP](https://coala.io/cep) of how the profiling
```
Comment (sims1253) [inline data/projects/debug_profile.md]: cEP should be merged at the end of preparation

File: data/projects/debug_profile.md [hunk 29]
```diff
@@ -1,26 +1,50 @@
...
   will be implemented.
 * The applicant is familiar with pdb.
 * The applicant is familiar with core coala codebase.
```
Comment (sims1253) [inline data/projects/debug_profile.md]: how to we test the above points?

File: data/projects/debug_profile.md [hunk 30]
```diff
@@ -1,26 +1,50 @@
...
 
-##### GSOC 2017 MIDTERM
+##### CODING PHASE 1
 
 * Pdb should be integrated completely into the coala core.
+* Documentation and tests for the above task should also have been completed.
```
Comment (sims1253) [inline data/projects/debug_profile.md]: should also have been -> are

File: data/projects/debug_profile.md [hunk 31]
```diff
@@ -1,26 +1,50 @@
...
+
+##### CODING PHASE 2
+
 * Settings of Bears can be accessed in the debug environment.
 * Documentation and tests for the above task should also have been completed.
```
Comment (sims1253) [inline data/projects/debug_profile.md]: same as above

File: data/projects/debug_profile.md [hunk 32]
```diff
@@ -1,26 +1,50 @@
...
+* The profile option to CLI is added.
```
Comment (sims1253) [inline data/projects/debug_profile.md]: is added to the CLI

File: data/projects/debug_profile.md [hunk 33]
```diff
@@ -1,26 +1,50 @@
...
 
-##### GSOC 2017 FINAL
+##### CODING PHASE 3
 
-* The profile option to CLI is added and merged.
-* The ability to inspect results produced by the bear is implemented.
+* The profile option to CLI is merged.
```
Comment (sims1253) [inline data/projects/debug_profile.md]: what does added mean if not merged?

File: data/projects/debug_profile.md [hunk 34]
```diff
@@ -1,26 +1,50 @@
...
+* The ability to inspect results produced by the Bear is implemented.
```
Comment (sims1253) [inline data/projects/debug_profile.md]: the bear -> Bears . No specific bear to implement this in was mentioned but rather we want this wo work in all bears right?

**Comment by sims1253** [comment]
ack d56196274cc4741aead0ea70ce881c7d27c6375c

File: data/projects/coala_cli.md [hunk 35]
```diff
@@ -1,24 +1,53 @@
...
+  been proposed by the participant.
+* Thorough testing of the implemented improvements is done and they are merged.
+* Documentation using illustrations of the improvements is merged.
 
-##### GSOC 2017 MIDTERM
+##### CODING PHASE 2
 
-* At least half of the points given in the cEP are done, merged, tested and properly documented.
+* Implemented, documented and tested at least 3 more improvements with coding
+  finished and merged.
```
Comment (Udayan12167) [inline data/projects/coala_cli.md]: So I feel we should leave it here and then judge the student based on his proposal. He will have more than a month to evaluate the code and write a proposal and then another month to rework the goal in his proposal. I think that would give us a clearer idea as well as the student's competence.

File: data/projects/debug_profile.md (hunk 27 above)
Comment (Udayan12167) [inline data/projects/debug_profile.md]: Couldn't think of a better motivation than a generic one.

File: data/projects/debug_profile.md (hunk 29 above)
Comment (Udayan12167) [inline data/projects/debug_profile.md]: I think merging the cEP means he would be familiar so changed the points.

**Comment by Udayan12167** [comment]
reack e70e8cf

File: data/projects/coala_cli.md (hunk 35 above)
Comment (sims1253) [inline data/projects/coala_cli.md]: I guess thats ok. Milestones are not set in stone and not realy important for the org application. Google likes us having them but doesn't care too much

File: data/projects/debug_profile.md (hunk 27 above)
Comment (sims1253) [inline data/projects/debug_profile.md]: Well its in there, just could be more visible :D
Why do we want to be able to debugg and profile?

File: data/projects/debug_profile.md (hunk 29 above)
Comment (sims1253) [inline data/projects/debug_profile.md]: agree

File: data/projects/coala_cli.md [hunk 36]
```diff
@@ -1,24 +1,51 @@
...
+In this project we are going to focus on the command line interface (CLI) of
+coala. The specific areas and points to focus on are:
+
+1. Remove most of the text from the action selection screen.
+2. Use letters instead of numbers for selecting actions.
+3. Allow chaining actions in a single input.
+4. Add a 'select action for all results' operator.
+5. Make the implementation of ApplyPatchAction log the applied patch.
+6. Reprompt after running actions only if the action says so.
+7. Set 'Do nothing' as the default action.
```
Comment (sims1253) [inline data/projects/coala_cli.md]: I think this is the case already

File: data/projects/coala_cli.md [hunk 37]
```diff
@@ -1,24 +1,51 @@
...
-Goal of this project is to finish the cEP and implement usability improvements
-defined in it.
+#### MILESTONES
 
-Ideally a candidate comes up with own ideas that are not sketched in the cEP
-yet!
+##### PREPARATION/BONDING
 
-#### MILESTONES
+* The [cEP](https://coala.io/cep) has been updated, finished and merged with
```
Comment (sims1253) [inline data/projects/coala_cli.md]: what cEP?`if no. 4, then link to it :D

File: data/projects/coala_cli.md [hunk 38]
```diff
@@ -1,24 +1,51 @@
...
+* A survey is done on the importance of each proposed change to the CLI.
+* Concept for implementation exists in the cEP.
 
-##### GSOC 2017 COMMUNITY BONDING
+##### CODING PHASE 1
 
-* The cEP has been updated, finished and merged.
-* The participant has a clear idea on how all parts of the cEP can be implemented.
+* Implemented at least 3 improvements from the above list plus whatever has
+  been proposed by the participant.
```
Comment (sims1253) [inline data/projects/coala_cli.md]: above list, including what has been proposed? sounds nicer

File: data/projects/improve_diff_handling.md [hunk 39]
```diff
@@ -1,24 +1,44 @@
-Bears should be able to offer more than just one possible patch
-for an issue. Users should be able to choose between different patches, e.g.
-for different possible spelling corrections.
-
-In addition bears should be able to offer templated patches, e.g. to offer
+One advantage of using coala is that some Bears offer automatically generated
+diffs to fix the problems they detect. This mechanism works smoothly as of now
+and even applies the proposed diff automatically to the problematic file if the
+user so desires. This is very helpful to the users. At the same time every
+issue in code can't have a single correct solution. The fix can and should
```
Comment (sims1253) [inline data/projects/improve_diff_handling.md]: I'D reword it to "At the same time, not every issue in the code can have just a single correct solution"

File: data/projects/coala_cli.md (hunk 36 above)
Comment (Udayan12167) [inline data/projects/coala_cli.md]: No I think it keeps prompting till you don't enter the 0 option.

**Comment by sims1253** [comment]
ack 4550f65eba46b209a37b129e010d5b86d98af00a

**Comment by sims1253** [comment]
reack ad094687b2ce7cd184966146e20b985a8ba186c4

**Comment by sims1253** [comment]
ack 9775056aa8539805be305dc616bbba876800d1aa

**Comment by sims1253** [comment]
ack 0f44a687f89868743bd11b39871dc5f3ec3b42de

**Comment by RohanVB** [comment]
Needs rebase

**Comment by Udayan12167** [comment]
reack 9916797

**Comment by Udayan12167** [comment]
reack 55a1c0a

**Comment by Udayan12167** [comment]
reack 6855ea1

**Comment by Udayan12167** [comment]
reack cfda0a0

**Comment by Udayan12167** [comment]
@rultor merge
